### DISCUSSION TITLE
Organizational structure of various Terraform modules

### DISCUSSION THREAD

**Discussion by RothAndrew** [defenseunicorns/delivery-aws-iac#166] [General]
Per the resolution of #146, we are going to move the modules to separate repos, but we haven't talked about what that will look like.

### What exists now?

* Right now we have a repo called [uds-infra](https://github.com/defenseunicorns/uds-infra) that is meant to act as the "umbrella" entrypoint to the various modules and zarf packages.
* We also have one module that has already been extracted into its own repo: the [tfstate-backend](https://github.com/defenseunicorns/terraform-aws-tfstate-backend)

### Known issues

* Having several independent repos will make things less DRY and make it harder to maintain things, unless we create automation to keep them in sync.

### Proposal

1. Pick the uds-infra repo back up. It has been sitting on ice for a little while now. Use it as the "one-stop-shop" that we will point people to as they start using our IaC stuff. It should be a sort of "directory" of what we offer, listing the various modules, root modules, zarf packages, and example repos
1. Create a GitHub template repo that will be used to create modules
1. Create automation that syncs the various module repos when changes to the template repo happens by automatically opening pull requests in the module repos
1. Create and document minimum requirements for testing and documentation for all modules
1. Start extracting modules, using the template repo to create new repos for each one, one by one.

### Notes

* I personally do not believe we should start extracting more modules until we have a template repo, the automation that will keep them in sync, and the testing and documentation reqs in place.

**Comment by RothAndrew** [comment]
cc @ntwkninja @brianrexrode @blancharda @wirewc

**Reply by wirewc** [reply]
Agreed. I will hopefully have an issue on a board to track to make a structure of modular repos and trying to address modularity at the zarf package level.

**Reply by ntwkninja** [reply]
Agree with the proposal.

Also, I have questions about the broader purpose of the current repo. 
- Should it be renamed and documented to be a specific implementation of iac? 
- Is the current implementation broad enough to be reusable to multiple use cases (i.e. does it transition to hosting multiple examples)?
- Do we anticipate that this repo will go beyond just aws?
- Are these examples or published versions of a root module that ties everything together?

cc: @runyontr

**Comment by wirewc** [comment]
After talking with @RothAndrew and @ntwkninja (with @justin-o12 and @zachariahmiller), a conclusion was made:
* https://github.com/defenseunicorns/uds/wiki/UDS-IaC#delivery-defined-mvp-must-haves
* https://github.com/defenseunicorns/uds/wiki/UDS-IaC#module-ownership-transition-plan
